src/Storages/RabbitMQ/StorageRabbitMQ.cpp

```diff
@@ -393,6 +676,53 @@ bool StorageRabbitMQ::streamToViews()
     std::atomic<bool> stub = {false};
     copyData(*in, *block_io.out, &stub);
 
+    /* Need to stop loop even if connection is ok, because sending ack() with loop running in another thread will lead to a lot of data
+     * races inside the library, but only in case any error occurs or connection is lost while ack is being sent
+     */
+    if (event_handler->loopRunning())
+    {
+        event_handler->updateLoopState(Loop::STOP);
+        looping_task->deactivate();
+    }
+
+    if (!event_handler->connectionRunning())
+    {
+        if (restoreConnection(true))
+        {
+            for (auto & stream : streams)
+                stream->as<RabbitMQBlockInputStream>()->updateChannel();
+
+        }
+        else
+        {
+            /// Reschedule if unable to connect to rabbitmq
+            return false;
+        }
+    }
+    else
+    {
+        heartbeat_task->deactivate();
+
+        /// Commit
+        for (auto & stream : streams)
+        {
+            if (!stream->as<RabbitMQBlockInputStream>()->sendAck())
```

Comment:
What will happen if we have already successfully inserted data to all views but one of streams got error sending `ack`? Is data duplication possible?

Reply:
hm, false is returned by the `sendAck` function in only two cases:
1) if connection failed. In this case all channels will be closed and will be unable to send `ack`. Also `ack` is made based on delivery tags, which is unique to channel, so if channels fail, those delivery tags will become invalid and there is no way to send specific `ack` from a different channel. Actually once the server realises that it has messages in a queue waiting for confirm from a channel which suddenly closed, it will immediately make those messages accessible to other consumers. So yes, in this case duplicates are inevitable. 
2) size of the sent frame (libraries's internal request interface) exceeds max frame - internal library error. This is more common for `message frames`, but not likely to happen to `ack frame` I suppose. So I do not believe it is likely to happen. But I can add one more check: if sendAck failed and connection still usable, retry to send `ack` and do not break from the loop in this case, but if frame size exceeded in the first time, then I believe for the second time nothing will change. Anyway I do not think that for `ack frame` this will ever happen

Reply:
Sorry for butting in, but this explanation would make an ideal comment for this code. This also goes for most things you ever have to explain to someone about the code.